src/audio/volume/volume.c

```diff
@@ -1155,6 +1155,58 @@ static vol_zc_func vol_get_zc_function(struct comp_dev *dev)
 	return NULL;
 }
 
+/**
+ * \brief Set volume frames alignment limit.
+ * \param[in,out] source Source structure.
+ * \param[in,out] sink Sink structure.
+ */
+#if XCHAL_HAVE_HIFI3 || XCHAL_HAVE_HIFI4
+static void volume_get_alignment(struct audio_stream *source,
+				 struct audio_stream *sink)
+{
+	uint16_t a;
+	uint16_t b;
+	uint16_t glc;
+	uint16_t limit_frames;
+	uint16_t count = 0;
+	/*the implement of S16 use 4-way intrinsic operation
+	 *frames should aligned with 4
+	 */
+	uint16_t alg_frames_limit = 4;
+	/*xtensa intrinsic aske forq 8-byte aligned*/
+	uint16_t intr_byte_limit = 8;
+	uint16_t frame_size = get_sample_bytes(source->frame_fmt) * source->channels;
+	/**calculate the largest common divisor*/
+	a = frame_size * alg_frames_limit;
+	b = intr_byte_limit;
+	while (a != b) {
+		if (a > b)
+			a = a - b;
+		else
+			b = b - a;
+	}
+	glc = a;
+	/**the least common multiple is frame_size * alg_frames_limit * intr_byte_limit
+	 *  / glc bytes, so the frames which aligned with both 4 frames and 8-byte is
+	 * frame_size * alg_frames_limit * intr_byte_limit / glc / frame_size.
+	 */
+	limit_frames = alg_frames_limit * intr_byte_limit / glc;
+
+	while (limit_frames > 1) {
+		limit_frames >>= 1;
+		count++;
+	}
+	source->frame_align_shift = count;
```

Comment:
maybe in sof/math/numbers.h or numbers.c would be better? Since I found there has already been a gcd function and these small functions are just for some logic calculatation.

Reply:
Yes I was thinking you might be able to use the existing gcd() function